### PR TITLE
Functors over effect values

### PR DESCRIPTION
**Pull request by googleson78**
I played around with this a bit to see how it would look.

What's peoples opinion on whether this actually brings any benefit  (i.e. some functions that use these classes? - I want to be able to write `ignore` which works over `Effunctors`, but alas, I don't think that is going to be possible :sob: )
vs
just writing them out (this has a clear benefit in my eyes - less surface area for new users)

Related to #165 and #168.

### CODE REVIEW

File: src/Polysemy/Internal/Effunctor.hs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+-- | Effects that are \"functors\" over their argument.
+--
+-- TODO: add better docs
+class Effunctor (e1 :: Type -> Effect) where
+  effmap
+    :: forall r a b c
+     . Member (e1 b) r
+    => (a -> b)
+    -> Sem (e1 a ': r) c
+    -> Sem r c
```
Comment (isovector) [inline src/Polysemy/Internal/Effunctor.hs]: I would expect this to be `Sem (e1 b ': r) c`

**Comment by isovector** [comment]
I'm in favor of this. Especially if we can TH-generate the functor instances!

**Comment by isovector** [comment]
@totbwf can your tactics stuff write these automatically for us yet? <3

File: src/Polysemy/Internal/Effunctor.hs (hunk 1 above)
Comment (googleson78) [reply src/Polysemy/Internal/Effunctor.hs]: I mirrored
```haskell
runOuputAsAnother :: Member (Output o2) r => (o1 -> o2) -> Sem (Output o1 ': r) a -> Sem r a
```
from [165](https://github.com/polysemy-research/polysemy/issues/165#issuecomment-509036665) for the type. What are the pros/cons of explicitly putting it in the front? (type inference/error messages/internal machinery/performance issues?)

Comment (isovector) [reply src/Polysemy/Internal/Effunctor.hs]: Good point. That's a very motivating example! Explicitly putting it at the front would mean we can't solve the [trouble with typed errors](https://www.parsonsmatt.org/2018/11/03/trouble_with_typed_errors.html)

**Comment by googleson78** [comment]
Do you have any ideas for other "functors" that would be common?

I will try to look into generating instances for these two with TH but:

0. I'm not 100% sure what the necessary and sufficient conditions for these are. (so I would appreciate any pointers/links)
1. I'm not experienced with TH, so it might take me a while to get it rolling when I'm done with 0.

File: src/Polysemy/Internal/Effunctor.hs (hunk 1 above)
Comment (Infinisil) [reply src/Polysemy/Internal/Effunctor.hs]: I seem to often write types like

```haskell
data Action r = forall s . Action
  { startingValue :: s
  , action        :: Int -> Sem (State s ': r) ()
  }

printAction :: Member Trace r => Action r
interactiveSetAction :: Member (Lift IO) r => Action r
```

And then I want `Action r` to be a `Semigroup` which runs both actions with their respective state:

```haskell
instance Semigroup (Action r) where
  (Action s a) <> (Action s' a') = Action (s, s') $ \value -> do
    mapStateFirst' (a value)
    mapStateSecond' (a' value)

mapStateFirst' :: Sem (State a ': r) x -> Sem (State (a, b) ': r) x
mapStateSecond' :: Sem (State b ': r) x -> Sem (State (a, b) ': r) x
```

For this case, these functions won't work:
```haskell
mapStateFirst :: Member (State (a, b)) r => Sem (State a ': r) x -> Sem r x
mapStateSecond :: Member (State (a, b)) r => Sem (State b ': r) x -> Sem r x
```

Failing with
<details>

```
    • Occurs check: cannot construct the infinite type:
        r ~ State (s, s1) : r
      Expected type: Sem (State (s, s1) : r) ()
        Actual type: Sem r ()
    • In a stmt of a 'do' block: mapStateFirst (a value)
      In the expression:
        do mapStateFirst (a value)
           mapStateSecond (a' value)
      In the second argument of ‘($)’, namely
        ‘\ value
           -> do mapStateFirst (a value)
                 mapStateSecond (a' value)’
    • Relevant bindings include
        a' :: Int -> Sem (State s1 : r) () (bound at tests/Main.hs:29:30)
        s' :: s1 (bound at tests/Main.hs:29:27)
        a :: Int -> Sem (State s : r) () (bound at tests/Main.hs:29:13)
        s :: s (bound at tests/Main.hs:29:11)
        (<>) :: Action r -> Action r -> Action r
          (bound at tests/Main.hs:29:16)
   |
30 |     mapStateFirst (a value)
   |     ^^^^^^^^^^^^^^^^^^^^^^^
```
</details>

I'm not sure if I'm designing something wrong here (should I not use `Semigroup`?), but if not, both of these function variants seem to have their use.

In my code I currently even use both variants. Usage is [here](https://github.com/Infinisil/arvy/blob/64f966cbd3afe0b7b87f480c94987f001c9a1be2/eval/Evaluation.hs#L23-L67), implementation of them is [here](https://github.com/Infinisil/arvy/blob/64f966cbd3afe0b7b87f480c94987f001c9a1be2/eval/Utils.hs#L38-L89).

**Comment by isovector** [comment]
For any GADT of the form `E i m a` we know that if `i` appears only in positive position, we should generate a `Effunctor E` instance. If `i` is only in negative position, then `Contraeffunctor`. Otherwise we don't have any instances for it.

**Comment by TheMatten** [comment]
@googleson78 I could do this as a separate PR if you want :slightly_smiling_face:
@isovector What about `InvEffunctor` in style of `Invariant` - could it have some practical use?

**Comment by googleson78** [comment]
> @googleson78 I could do this as a separate PR if you want

Let me have some fun too! (unless I'm taking too long)

File: src/Polysemy/Internal/Effunctor.hs (hunk 1 above)
Comment (googleson78) [reply src/Polysemy/Internal/Effunctor.hs]: I haven't gone too deep into your example, but why not write `action` as `Member (State s) r => Int -> Sem r ()` instead?

Comment (TheMatten) [reply src/Polysemy/Internal/Effunctor.hs]: He later tries to map `State` in this action, which requires him to remove original and add a modified one - `Member` constraint is not enough for that.

**Comment by isovector** [comment]
What's the state of this?

**Comment by googleson78** [comment]
0. There needs to be a decision on which (if not both) form of `effmap` should be in the type class (return-type polymorphic or not). - Don't know.
1. TH autogeneration needs implementing. - I haven't had the time recently. If anyone else wants to, go ahead!

**Comment by TheMatten** [comment]
@googleson78 I may find some time for 1.

**Comment by isovector** [comment]
@TheMatten and I spent some time thinking about this, but it unfortunately doesn't play very nicely with higher order effects.

**Pull request closed without merging**